### PR TITLE
Rewrote SSE geometry code

### LINKED ISSUE #1010: Refactor geometry kernels
The core geometry functions (`compute_distances`, `compute_angles`, etc) are written in a style in C that is not very inviting. The preprocessor is used in a kind of dirty way to get both the periodic and non-periodic versions of the kernels compiled, and the explicit use of SSE intrinsics is a little challenging.

It should probably be refactored into a more sensible C++ using [OpenMM's](https://github.com/pandegroup/openmm/blob/eb1293f52263e49536cb2004799fe9cd293a0712/openmmapi/include/openmm/internal/vectorize_sse.h) sse wrappers to simplify the visual look of all of the SSE code.

### PR DESCRIPTION
**Pull request by peastman**
I've rewritten all the optimized SSE based code, replacing the direct calls to intrinsics with wrapper classes.  This fixes #1010.  The main goal of this was to make the code easier to understand and maintain, but in the process I fixed several bugs and made other improvements:

- There was a bug in the implementation of dssp that caused it to sometimes misclassify residues.  See #1070.

- For triclinic boxes, `compute_distances()` and `compute_displacements()` would sometimes identify the wrong periodic copy (not the closest one).  See #1066.  I've improved that.  I'm still not 100% certain it is always guaranteed to find the closest copy, but it's certainly better than it was.  This may also address #1019, but I haven't tested that.

- If the box vectors changed over the trajectory, `compute_neighbors()` would use the vectors from the first frame for every frame.  I fixed that.

- `compute_angles()` and `compute_dihedrals()` didn't have optimized code for triclinic boxes, so in that case they would fall back to the very slow Python version.  The SSE code for them now supports triclinic boxes.

### CODE REVIEW

**Comment by mpharrigan** [comment]
Wow! Nice work

File: mdtraj/geometry/src/geometry.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+ * not. Most of the code between these versions is shared, so I've written
+ * the parts of the two functions which are different using #ifdefs. So we
+ * just include these files _twice_ here, toggling the variable that controls
+ * the ifdef.
+ *
+ * Note that these kernel files are really not capable of being compiled
+ * independently -- they're not header files at all -- and they're really just
+ * meant to be #included here.
+ **/
+#undef COMPILE_WITH_PERIODIC_BOUNDARY_CONDITIONS
```
Comment (marscher) [inline mdtraj/geometry/src/geometry.cpp]: Nice hack. Does this increase maintainability as you stated?

**Comment by marscher** [comment]
This is major improvement! Thank you

File: mdtraj/geometry/src/geometry.cpp (hunk 1 above)
Comment (peastman) [inline mdtraj/geometry/src/geometry.cpp]: It's not my hack!  I just copied that directly from the old geometry.c file.  So it still works the same way as before, just with different code in the included files.  (Personally, I don't find it all that hacky, but whatever.)

Comment (swails) [inline mdtraj/geometry/src/geometry.cpp]: > (Personally, I don't find it all that hacky, but whatever.)

I remember thinking exactly the same thing when I saw it in `geometry.c` :smiley:.  This is a fairly common technique I've seen used in many places including several codes in Amber as well as OpenMM (in a sense, anyway).

Comment (rmcgibbo) [inline mdtraj/geometry/src/geometry.cpp]: I did write it, but I guess I'm the only one who finds it ugly. :)

**Comment by rmcgibbo** [comment]
Big +1 from me. Thanks, @peastman!

**Pull request merged**